File: algorithms/algorithms-cplusplus/leetcode/Question_0722_Remove_Comments.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<string> removeComments(vector<string>& source) {
    vector<string> res;
    string s;
    bool comment = false;
    for (string& l : source) {
      for (int i = 0; i < l.size(); i++) {
        if(!comment && i + 1 < l.size() && l[i] == '/' && l[i + 1] == '/') {
          break;
        }
        if (!comment && i + 1 < l.size() && l[i] == '/' && l[i + 1] == '*') {
          comment = true;
          i++;
        } else if (comment && i + 1 < l.size() && l[i] == '*' && l[i + 1] == '/') {
          comment = false;
          i++;
        } else if (!comment) {
          s.push_back(l[i]);
        }
      }
      if(!comment && s.size()) {
        res.push_back(s);
        s.clear();
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0722_Remove_Comments.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0722_Remove_Comments {
    public List<String> removeComments(String[] source) {
        List<String> res = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        boolean comment = false;
        for (String s : source) {
            char[] arr = s.toCharArray();
            for (int i = 0; i < arr.length; i++) {
                if (!comment && i + 1 < arr.length && arr[i] == '/' && arr[i + 1] == '/') {
                    break;
                }
                if (!comment && i + 1 < arr.length && arr[i] == '/' && arr[i + 1] == '*') {
                    comment = true;
                    i++;
                } else if (comment && i + 1 < arr.length && arr[i] == '*' && arr[i + 1] == '/') {
                    comment = false;
                    i++;
                } else if (!comment) {
                    sb.append(arr[i]);
                }
            }
            if (!comment && sb.length() > 0) {
                res.add(sb.toString());
                sb = new StringBuilder();
            }
        }
        return res;
    }
}
